VariantEncoding.md

```diff
@@ -416,14 +444,36 @@ Field names are case-sensitive.
 Field names are required to be unique for each object.
 It is an error for an object to contain two fields with the same name, whether or not they have distinct dictionary IDs.
 
-# Versions and extensions
+## Versions and extensions
 
 An implementation is not expected to parse a Variant value whose metadata version is higher than the version supported by the implementation.
 However, new types may be added to the specification without incrementing the version ID.
 In such a situation, an implementation should be able to read the rest of the Variant value if desired.
 
-# Shredding
+## Shredding
 
 A single Variant object may have poor read performance when only a small subset of fields are needed.
 A better approach is to create separate columns for individual fields, referred to as shredding or subcolumnarization.
 [VariantShredding.md](VariantShredding.md) describes the Variant shredding specification in Parquet.
+
+## Conversion to JSON
+
+Values stored in the Variant encoding are a superset of JSON values.
+For example, a Variant value can be a date that has no equivalent type in JSON.
+To maximize compatibility with readers that can process JSON but not Variant, the following conversions should be used when producing JSON from a Variant:
+
+| Variant type  | JSON type | Representation requirements                              | Example                              |
+|---------------|-----------|----------------------------------------------------------|--------------------------------------|
+| Null type     | null      | `null`                                                   | `null`                               |
+| Boolean       | boolean   | `true` or `false`                                        | `true`                               |
+| Exact Numeric | number    | Digits in fraction must match scale, no exponent         | `34`, 34.00                          |
```

Comment:
For exact numerics, we should allow truncating trailing zeros. For example, `int8` value `1` and `decimal(5,2)` value `100` can both be represented as a JSON value `1`.

Also, should the example be quoted to stay consistent?
```suggestion
| Exact Numeric | number    | Digits in fraction must match scale, no exponent         | `34`, `34.00`                          |
```

Reply:
I think the intent of considering `Exact Numeric` to be a single logical type is that we consider the `int8` value `1` to be logically equivalent to `decimal(5,2)` with unscaled value `100`. If that's the case, I think we'd want the produced JSON to be the same for both (probably `1` in both cases), and not recommend having the fraction match the scale.

Reply:
@gene-db, @cashmand, these are concerns for the engine layer, not for storage. If Spark wants to automatically coerce between types that's fine, but the compromise that we talked about a couple months ago was to leave this out of the shredding spec and delegate the behavior to engines. Storage should always produce the data that was stored, without modification.

Reply:
Yes, the engine should be the one concerned with changing types.

However, my original question was about this JSON representation wording. Currently, the `Representation requirements` for an `Exact Numeric` says the `Digits in fraction must match scale`. However, because the `Exact Numeric` is considered a logical type, the value `1` could be stored in the Variant as `int8` 1 or `decimal(5,2)` 100. Both of those would be the same numeric value, so we should allow truncating trailing zeros in the JSON representation, instead of requiring that the digits in the fraction match the scale.

Reply:
@rdblue I am confused with this JSON chart then. If we are talking about "storage implementation", then are you expecting there is a "storage implementation" that is converting variant values to JSON? When will storage convert a variant value to a JSON string?

I originally thought this chart was trying to say, "When an engine wants to convert a variant value to a JSON string, here are the rules". Therefore, we should allow engines to cast integral decimals to integers before converting to JSON, as you already mentioned in your previous comment.

Reply:
I think I agree with @gene-db on this.  I think any json representation that has semantically the same meaning in JSON should be allowed.  Translation to JSON is inherently lossy and I think trying to match semantics will be more error prone then it is worth (i.e. it  should be a non-goal to expect it to be able to reconstruct the exact same variant from the proposed JSON representation).

Reply:
> When an engine wants to convert a variant value to a JSON string, here are the rules

Yes, this is correct. We want a clear way to convert to a JSON string. However, the normalization needs to happen first. We don't want to specify that the JSON must be any more lossy than it already is.

Why would we require an engine to produce a normalized value?

Reply:
> Why would we require an engine to produce a normalized value?

At least for me, I don't think it is about "requiring" and engine to produce a normalized value first.  I think if an engine is reading variant and converting it to JSON, it is possibly doing so through an internal representation so it can still apply operators on top of the JSON value and possibly even storing it as an internal representation.  Conversion to a string is really only an end-user visible thing.  So when I read this it seems to be requiring an engine to NOT normalize which could be hard to implement for some engines.